### PR TITLE
PropertyTreeEditor | Change container set behavior to use destination type and circumvent issues with basic_string

### LINKED ISSUE #10464: PostFX Layer component camera tag lists can't add container item string value with Editor Python Bindings
**Describe the bug**
When automating PostFX Layer component I'm attempting to add a string value to the container item 
`'Controller|Configuration|Select Camera Tags Only'`
`AZStd::vector<AZStd::string> m_cameraTags`
and
`'Controller|Configuration|Excluded Camera Tags'`
`AZStd::vector<AZStd::string> m_exclusionTags`

I was able to get a value saved in a level on an entity through editor then loaded by automation. for that manually created and saved tag I'm able to read the string value that exists. I was unable to update the existing value.
When a new PostFX Layer component is created and a tag string is either appended to the container or the value is added to specified index 0 key; in both cases the container count reaches 1 indicating there is a container item but the get of the container item value returns empty string
```
        postfx_layer_component.append_container_item('Controller|Configuration|Select Camera Tags Only', value='yes')
        item = postfx_layer_component.get_container_item('Controller|Configuration|Select Camera Tags Only', key=0)
        Report.info(f"'{item}' {type(item)} {postfx_layer_component.get_container_count('Controller|Configuration|Select Camera Tags Only')}")
```
returns
` Info: '' <class 'str'> 1`

when an existing item is read from a saved level that was manually created the value can be read but not updated
`Info: 'yes' <class 'str'> 1`

**Assets required**
[!Important!] Please provide sample assets needed to reproduce the issue, either as an [attachment](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files) or as a link to a public asset in Github.

**Steps to reproduce**
1. using the existing test automation script Atom/tests/hydra_AtomEditorComponents_PostFXLayerAdded.py
2. append the following lines to the script after component creation
```
        postfx_layer_component.append_container_item('Controller|Configuration|Select Camera Tags Only', value='yes')
        item = postfx_layer_component.get_container_item('Controller|Configuration|Select Camera Tags Only', key=0)
        Report.info(f"'{item}' {type(item)} {postfx_layer_component.get_container_count('Controller|Configuration|Select Camera Tags Only')}")
```
3. execute the script in editor using the console command
`pyRunFile ../../Gem/PythonTests/Atom/tests/hydra_AtomEditorComponents_PostFXLayerAdded.py`

**Expected behavior**
console output includes
`Info: 'yes' <class 'str'> 1`

**Actual behavior**
console output includes the line
` Info: '' <class 'str'> 1`

**Screenshots/Video**
If applicable, add screenshots and/or a video to help explain your problem.

**Found in Branch**
commit d1d9d7c76d4d6cffcadce6fc9091c53169a36e12 (HEAD -> development

**Desktop/Device (please complete the following information):**
windows

**Additional context**
Add any other context about the problem here
postfxLayer

### PR DESCRIPTION
**Pull request by AMZN-daimini**
## What does this PR do?
Changes the behavior of the PTE's container item setters to use the destination's type when cloning in place. Note that SetProperty in the PTE already uses the destination's type to determine how to convert the Any value passed to the function, but the container API that was added separately did not.

Fixes #10464 

## How was this PR tested?

This script was used to test the behavior. It was run via the Phyton Scripts interface inside the Editor:
```
import azlmbr.legacy.general as general
import azlmbr.bus as bus
import azlmbr.editor as editor
import azlmbr.entity as entity

def find_entity_by_name(entity_name):
    """
    Gets an entity ID from the entity with the given entity_name
    :param entity_name: String of entity name to search for
    :return entity ID
    """
    search_filter = entity.SearchFilter()
    search_filter.names = [entity_name]
    matching_entity_list = entity.SearchBus(bus.Broadcast, 'SearchEntities', search_filter)
    if matching_entity_list:
        matching_entity = matching_entity_list[0]
        if matching_entity.IsValid():
            print(f'{entity_name} entity found with ID {matching_entity.ToString()}')
            return matching_entity
    else:
        return matching_entity_list

entityId = find_entity_by_name("Entity1")

type_ids_list = editor.EditorComponentAPIBus(bus.Broadcast, 'FindComponentTypeIdsByEntityType', ["PostFX Layer"], entity.EntityType().Game)
postFxLayerTypeId = type_ids_list[0]

postFxLayerComponentOutcome = editor.EditorComponentAPIBus(bus.Broadcast, 'GetComponentOfType', entityId, postFxLayerTypeId)
postFxLayerComponent = postFxLayerComponentOutcome.GetValue()

build_prop_tree_outcome = editor.EditorComponentAPIBus(bus.Broadcast, "BuildComponentPropertyTreeEditor", postFxLayerComponent)
property_tree_editor = build_prop_tree_outcome.GetValue()

property_tree_editor.append_container_item('Controller|Configuration|Select Camera Tags Only', 'yes')
```

Before the change, this error would be printed:
```
[Error] (Serialize) - Element with class ID '{7114E998-A8B4-519B-9342-A86D1587B4F7}' is not registered with the serializer!

=== Serialize stack ===
[ Address: 000001F880719500 Uuid: {7114E998-A8B4-519B-9342-A86D1587B4F7} ]
```
(`{7114E998-A8B4-519B-9342-A86D1587B4F7}` is the typeId for the basic string)

After the change, no error is printed and the Component's value is updated as expected
image

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp [hunk 1]
```diff
@@ -601,7 +600,7 @@ namespace AzToolsFramework
                 if (keyAddress && valueAddress)
                 {
                     m_serializeContext->CloneObjectInplace(keyAddress, AZStd::any_cast<void>(&key), key.type());
-                    m_serializeContext->CloneObjectInplace(valueAddress, AZStd::any_cast<void>(&value), value.type());
+                    m_serializeContext->CloneObjectInplace(valueAddress, AZStd::any_cast<void>(&value), data.value().m_valueElement->m_typeId);
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp:603]: We should verify that the source value is the type or derived from the destination type before copying.
We should be able to use the `data.value().m_valueElement.m_azRtti` to check if the type associated with the AZStd::any is derived from the the destination by using the `IRttiHelper::IsTypeOf` function

**Review by cgalvan — APPROVED** [review]
+1 to @lumberyard-employee-dm suggestion

File: Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp (hunk 1 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp:603]: Thanks, will do!

Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp:603]: Alright, I've been playing around with this but it turns out `value.type()` for strings will return the typeId `basic string)`, which isn't registered to the serialize context. So trying to use functions like `IsTypeOf`, or `m_serializeContext->FindClassData` etc. will not work with this. The conversion works, though.

So far, i've assumed that basic_string is intentionally not reflected, but I'm not sure if that's the case?

Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp:603]: @lumberyard-employee-dm is there an alternative to IsTypeOf that can let us identify conversions in this case?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzToolsFramework/AzToolsFramework/PropertyTreeEditor/PropertyTreeEditor.cpp:603]: The basic_string type is registered with the SerializeContext implicitly when it is reflected as ->Field in the SerializeContext reflection.
So the typeid for is registered with the SerializeContext through the [SerializeGenericTypeInfo](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl#L3291-L3354) system

**Review by gadams3 — APPROVED** [review]

**Pull request closed without merging**